packages/template-retail-react-app/app/hooks/use-auth-modal.js

```diff
@@ -38,69 +44,91 @@ export const AuthModal = ({
+    )
     const navigate = useNavigation()
     const [currentView, setCurrentView] = useState(initialView)
     const form = useForm()
     const submittedEmail = useRef()
     const toast = useToast()
+    const login = useShopperLoginHelper(ShopperLoginHelpers.LoginRegisteredUserB2C)
+    const register = useShopperLoginHelper(ShopperLoginHelpers.Register)
 
     const submitForm = async (data) => {
         form.clearErrors()
 
-        return {
-            login: handleLogin,
-            register: handleRegister,
-            password: handleResetPassword
-        }[currentView](data)
-    }
+        // mutation is cached by default?
+        // TODO: should we disable mutation cache by default?
+        login.reset()
+        register.reset()
 
-    const handleLogin = async (data) => {
-        try {
-            await customer.login(data)
-        } catch (error) {
-            const message = /invalid credentials/i.test(error.message)
-                ? formatMessage({
-                      defaultMessage:
-                          "Something's not right with your email or password. Try again.",
-                      id: 'auth_modal.error.incorrect_email_or_password'
-                  })
-                : formatMessage(API_ERROR_MESSAGE)
-            form.setError('global', {type: 'manual', message})
-        }
-    }
-
-    const handleRegister = async (data) => {
-        try {
-            await customer.registerCustomer(data)
+        const onLoginSuccess = () => {
             navigate('/account')
-        } catch (error) {
-            form.setError('global', {type: 'manual', message: error.message})
         }
-    }
 
-    const handleResetPassword = async ({email}) => {
-        try {
-            await customer.getResetPasswordToken(email)
-            // Execute action to be perfromed on successful passoword reset
-            await onPasswordResetSuccess()
-            submittedEmail.current = email
-        } catch (error) {
-            form.setError('global', {type: 'manual', message: error.message})
-        }
+        return {
+            login: (data) =>
+                login.mutateAsync(
+                    {username: data.email, password: data.password},
+                    {
+                        onSuccess: onLoginSuccess,
+                        onError: (error) => {
+                            const message = /Unauthorized/i.test(error.message)
+                                ? formatMessage({
+                                      defaultMessage:
+                                          "Something's not right with your email or password. Try again.",
+                                      id: 'auth_modal.error.incorrect_email_or_password'
+                                  })
+                                : formatMessage(API_ERROR_MESSAGE)
+                            form.setError('global', {type: 'manual', message})
+                        }
+                    }
+                ),
```

Comment:
**Observation**

The onSuccess and onError handlers are saving out but a lot for these kinda of mutation actions. We are using a synchronous action for the submitHandler and relying on the onSuccess and onError handlers previously mentioned to manually set the state of the form. 

I'm not saying that this is the wrong way to do it, but it seem as tho the hook style mutations is harder to use in these types of cases. I wonder if instead of doing something like this, we have a utilty to get the underlying api async method from the hook, and we can call that in an async way, and return the expected result as defined by the useForm API. 

Side note.. is the form loading state going to be incorrect in this case? Since the function is sync, it'll only be loading momentarily whilst the network request still hasn't returned.

Reply:
> We are using a synchronous action for the submitHandler

This is not true, `mutateAsync` returns the promise that is the underlying method.

Reply:
That being said, we could completely re-write this to be more hook-y. To not use onSuccess/onError. For example

```
const component = () => {

    const login = useShopperLoginHelper('login')
    const submitHandler = login.mutate

   return <form error={login.error}></form>
}
```

But I don't see a whole lot of benefit and more risk to break the existing patterns. The existing patterns does the job pretty well.

Reply:
The handler is not being defined as async.. so even though you are calling a asyn function, the handler is executing to completion 

```
            login: (data) =>
                login.mutateAsync(
                    {username: data.email, password: data.password},
                    {

Reply:
BTW the comment about our pattern isn't blocking.

Reply:
Maybe i'm missing something basic but why do we need to define this function as async if the only thing it does is to return a promise, why is executing to completion a problem?

My understanding is that async/await is just a syntactic sugar to make a function return a promise without actually returning a promise.

Reply:
oops, i think my reply is out-dated. gonna resolve the conversation